dask/dataframe/io/parquet/utils.py

```diff
@@ -0,0 +1,296 @@
+
+    @staticmethod
+    def read_metadata(
+        fs, fs_token, paths, categories=None, index=None, gather_statistics=None
+    ):
+        """ Gather metadata about a Parquet Dataset to prepare for a read
+
+        This function is called once in the user's Python session to gather
+        important metadata about the parquet dataset.
+
+        Parameters
+        ----------
+        fs: FileSystem
+        fs_token: ??
+        paths; List[str]
+            A list of paths to files (or their equivalents)
+        categories:
+        index: str  # TODO: maybe remove this?
+            A suggested column to use as the index, if provided by the user
+        gather_statistics: bool
+            Whether or not to gather statistics data.  If ``None`` we only
+            gather statistics data if there is a single .metadata file to
+            cheaply query
+
+        Returns
+        -------
+        meta: pandas.DataFrame
+            An empty DataFrame object to use for metadata.
+            Should have appropriate column names and dtypes but need not have
+            any actual data
+        statistics: Optional[List[Dict]]
+            Either none, if no statistics were found, or a list of dictionaries
+            of statistics data, one dict for every partition (see the next
+            return value).  The statistics should look like the following:
+
+                [
+                    {'num-rows': 1000, 'columns': [
+                        {'name': 'id', 'min': 0, 'max': 100, 'null-count': 0},
+                        {'name': 'value', 'min': 0.0, 'max': 1.0, 'null-count': 5},
+                        ]},  # TODO: we might want to rethink this oranization
+                    ...
+                ]
+        parts: List[object]
+            A list of objects to be passed to ``Engine.read_partition``.
+            Each object should represent a row group of data.
+            We don't care about the type of this object, as long as the
+            read_partition function knows how to interpret it.
+        """
+        raise NotImplementedError()
+
+    @staticmethod
+    def read_partition(fs, piece, columns, partitions, categories, *args):
+        """ Read a single piece of a Parquet dataset into a Pandas DataFrame
+
+        This function is called many times in individual tasks
+
+        Parameters
+        ----------
+        fs: FileSystem
+        piece: object
+            This is some token that is returned by Engine.read_metadata.
+            Typically it represents a row group in a Parquet dataset
+        columns: List[str]
+            List of column names to pull out of that row group
+        partitions:
+        categories:
+
+        Returns
+        -------
+        A Pandas DataFrame
+        """
+        raise NotImplementedError()
+
+    @staticmethod
+    def create_metadata(df, fs, path, append=False, **kwargs):
+        """Create engine-specific metadata instance for this dataset
```

Comment:
Looking at this function again, some thoughts/questions:

1.  I think that we should change name to avoid the term `metadata`, which is already somewhat overloaded.  Perhaps arguments or parameters rather than metadata?
2.  I'm curious about what sorts of engine-specific parameters come up here?  If it's possible to handle these in some other way I'd be inclined to remove this.  When I say some other way I mean ...
    1. I would think that some of these options, like filenames, wouldn't be the responsibility of the engine.  At first glance I would expect this to be lifted to the core.
    2. Perhaps some of these policies should be handled in the `write_partition` function instead?

cc @martindurant

Reply:
I have no particular love for the "metadata" name here. Seemed more specific than "container", "instance"... Maybe call it something like "initialise"?

I think that the process particularly around appending is specific enough in the engines that it would be difficult to pull it out of here and generalise, make the method smaller or go away altogether. 
I suspect that the object encoding, divisions and filenames are also things that need to be decided beforehand and depend on the particular engine, but I am less sure about these.

Reply:
> I suspect that the object encoding, divisions and filenames are also things that need to be decided beforehand and depend on the particular engine, but I am less sure about these.

I agree that they should be decided beforehand, but I would be surprised to find that they should be engine specific.  I think that ideally I would get a dataset with the same divisions and filenames regardless of which engine I chose.  

> I think that the process particularly around appending is specific enough in the engines that it would be difficult to pull it out of here and generalise, make the method smaller or go away altogether.

How would fastparquet and arrow handle things differently here?  What sorts of engine-specific issues would arise?  Do these need to be handled ahead of time or can they be handled within the write_partition function?

My (probably naive) guess was that when appending we would look at the metadata of the existing file, check statistics (maybe?), and then create a bunch of new filenames that start at n + 1 and map `Engine.write_partition` over those with the new partitions.